File: test_vs/02_concurrency.py
"""Benchmark: Concurrent Message Passing — Python asyncio baseline."""

import asyncio
import time


async def worker(inbox: asyncio.Queue, outbox: asyncio.Queue, stop_event: asyncio.Event):
    while not stop_event.is_set():
        try:
            val = await asyncio.wait_for(inbox.get(), timeout=0.1)
            await outbox.put(val * 2 + 1)
        except asyncio.TimeoutError:
            continue


async def main():
    inbox: asyncio.Queue[int] = asyncio.Queue(maxsize=1024)
    outbox: asyncio.Queue[int] = asyncio.Queue(maxsize=1024)
    stop = asyncio.Event()

    task = asyncio.create_task(worker(inbox, outbox, stop))

    start = time.perf_counter()
    total = 0
    for i in range(10_000):
        await inbox.put(i)
        result = await outbox.get()
        total += result
    elapsed = time.perf_counter() - start

    stop.set()
    await task

    print(f"sum = {total}")
    print(f"Time: {elapsed:.4f}s")


if __name__ == "__main__":
    asyncio.run(main())

File: benchmarks/cross_language/02_concurrency.go
// Benchmark: Concurrent Message Passing -- Go goroutines + channels

package main

import (
	"fmt"
	"runtime"
	"time"
)

func worker(inbox <-chan int, outbox chan<- int) {
	for val := range inbox {
		outbox <- val*2 + 1
	}
	close(outbox)
}

func main() {
	inbox := make(chan int, 1024)
	outbox := make(chan int, 1024)

	runtime.GC()
	var memBefore runtime.MemStats
	runtime.ReadMemStats(&memBefore)

	go worker(inbox, outbox)

	start := time.Now()
	sum := 0
	for i := 0; i < 10000; i++ {
		inbox <- i
		result := <-outbox
		sum += result
	}
	close(inbox)
	elapsed := time.Since(start)

	var memAfter runtime.MemStats
	runtime.ReadMemStats(&memAfter)

	fmt.Printf("sum = %d\n", sum)
	fmt.Printf("__BENCH_METRICS__\n")
	fmt.Printf("wall_time_s=%.6f\n", elapsed.Seconds())
	fmt.Printf("peak_memory_kb=%.1f\n", float64(memAfter.TotalAlloc-memBefore.TotalAlloc)/1024.0)
}
